cpg_pipes/providers/inputs.py

```diff
@@ -0,0 +1,417 @@
+                        sample=s,
+                        fam_id=fam_id,
+                        dad=sample_by_participant_id.get(pat_id),
+                        mom=sample_by_participant_id.get(mat_id),
+                        sex=Sex.parse(sex),
+                        phenotype=phenotype or '0',
+                    )
+
+    def _get_entries(
+        self,
+        dataset: Dataset | None = None,
+        skip_samples: list[str] | None = None,
+        only_samples: list[str] | None = None,
+        skip_datasets: list[str] | None = None,
+    ) -> list[dict]:
+        """
+        Helper method to get and filter entries.
+        """
+        entries = self.get_entries(dataset)
+        entries = self._filter_samples(
+            entries=entries,
+            skip_samples=skip_samples,
+            only_samples=only_samples,
+            skip_datasets=skip_datasets,
+        )
+        return entries
+
+    def _filter_samples(
+        self,
+        entries: list[dict[str, str]],
+        skip_samples: list[str] | None = None,
+        only_samples: list[str] | None = None,
+        skip_datasets: list[str] | None = None,
+    ) -> list[dict[str, str]]:
+        """
+        Apply the only_samples and skip_samples filters.
+        """
+        filtered_entries = []
+        for entry in entries:
+            cpgid = self.get_sample_id(entry)
+            extid = self.get_external_id(entry)
+            ds = self.get_dataset_name(entry)
+            if skip_datasets and ds and ds in skip_datasets:
+                logger.info(
+                    f'Skipping sample (the dataset is skipped): {ds}|{cpgid}|{extid}'
+                )
+                continue
+            if only_samples:
+                if cpgid in only_samples or extid in only_samples:
+                    logger.info(f'Picking sample: {ds}|{cpgid}|{extid}')
+                else:
+                    continue
+            if skip_samples:
+                if cpgid in skip_samples or extid in skip_samples:
+                    logger.info(f'Skiping sample: {ds}|{cpgid}|{extid}')
+                    continue
+            filtered_entries.append(entry)
+        return filtered_entries
+
+    def _add_sample(self, dataset: Dataset, entry: dict):
+        """
+        Helper method to create a sample in a dataset.
+        """
+        entry = dict(entry)  # making a copy, so methods can call .pop()
+        dataset.add_sample(
+            id=str(self.get_sample_id(entry)),
+            external_id=str(self.get_external_id(entry)),
+            participant_id=self.get_participant_id(entry),
+            sex=self.get_participant_sex(entry),
+            sequencing_type=self.get_sequencing_type(entry),
+            meta=self.get_sample_meta(entry),
+        )
+
+
+class FieldMap(Enum):
+    """
+    CSV field names.
+    """
+
+    dataset = 'dataset'
+    sample = 'sample'
+    external_id = 'external_id'
+    participant_id = 'participant_id'
+    fqs_r1 = 'fqs_r1'
```

Comment:
Why pass them like this, and not fq_pairs, and just chunk by two? Is this generally easier to produce?

Reply:
Basically, lack of delimiters. We should support CSV and TSV files, and we use `|` to separate FASTQs from multiple lanes. So we need yet another delimiter to separate pairs, and it would get really confusing at this point. I guess I could borrow your parser code from sample-metadata, but didn't want to invest too much at the CSV provider, as we are only going to be using the sample-metadata input provider.